packages/typespec-client-generator-core/src/public-utils.ts

```diff
@@ -698,3 +707,61 @@ export function isAzureCoreModel(t: SdkType): boolean {
 export function isPagedResultModel(context: TCGCContext, t: SdkType): boolean {
   return context.__pagedResultSet.has(t);
 }
+
+/**
+ * Find corresponding http parameter list for a client initialization or service method parameter or the property of that parameter.
+ *
+ * @param method
+ * @param param
+ * @returns
+ */
+export function getHttpOperationParameter(
+  method: SdkServiceMethod<SdkHttpOperation>,
+  param: SdkModelPropertyType,
+):
+  | SdkPathParameter
+  | SdkQueryParameter
+  | SdkHeaderParameter
+  | SdkCookieParameter
+  | SdkBodyParameter
+  | undefined {
+  const operation = method.operation;
+  const queue: SdkModelPropertyType[] = [param];
+  const visited: Set<SdkModelType> = new Set();
+  // BFS to find the corresponding http parameter.
+  // An http parameter will be mapped to a method/client parameter, several method/client parameters (body spread case), or one property of a method property (metadata on property case).
+  // So, when we try to find which http parameter a method parameter corresponds to, we compare the `correspondingMethodParams` list directly.
+  // When we try to find which http parameter a property corresponds to, we need to consider the following cases:
+  // 1. The property itself is an http parameter.
+  // 2. The property is a model, and the properties of the model are http parameters, recursively.
+  // 3. The property is a model, and the base model of the model has http parameters, recursively.
+  // When we find one, we return directly since a method/client parameter or property could only be used in one http parameter.
```

Comment:
Is this assert valid for below case?

```
      model Params {
        @query foo: string;
        @query bar: string;
      }
      op func(...Params): void;
      `,
      `
      namespace MyCustomizations;
      op func(params: MyService.Params): void;
      @@override(MyService.func, MyCustomizations.func);
```

Reply:
i think so. for `params`, its corresponding http parameters is `undefined`. you need to get them with `foo` and `bar` property. see my last test.

Reply:
Ok, so a method parameter would map to at most 1 parameter, and strictly an HTTP parameter (body param or query param or header param or `undefined`, but not to property in body).

It is somewhat unexpected that one can go to the properties inside the model and call `getHttpOperationParameter` on it. Let's see what other reviewers think.

Reply:
a little bit force and back for this. just want to make it simple. current spec does not contain such mix usage at least.

Reply:
One concern is, if I call the method, and find the parameter maps to a kind=body, I seems won't be able to tell at this state, whether the parameter is the body, or it actually be a property within that body?

Reply:
currently, it will tell you this parameter is used as body. but it could not tell you the property of this parameter is also used as a query patameter.

Reply:
What I mean is,
```
model Model { key: string }

op myOp(@header h: string, @query q: string, @path p: string, @body body: Model): void;
```
call on "body" method param would map to `body` body param of the `Model`.

And
```
op myOp(@header h: string, @query q: string, @path p: string, ...Model): void;
```
call on "key" method param would also map to `body` body param of the `Model`.
```

Maybe it is not that bad, maybe we can use the mapping from the other side to determine...

Reply:
i got your point. you mean there is no way to tell how to map the exact value. then you are right, tcgc suggest to map from http parameter to method parameter.

Reply:
I tend to agree with weidong that we should provide such information whether a parameter is from the body property or from the body itself otherwise I don't know how to handle the serialization logic for spread body model with only one property inside it ?

Reply:
then maybe i need to give you a segment array?

Reply:
I think that works. And it will also works if we want to support group a few parameters into one parameter like @weidongxu-microsoft proposed before?